src/Application/Tools/Algorithm/FastGrowCut.h

```diff
@@ -0,0 +1,89 @@
+/*
+   For more information, please see: http://software.sci.utah.edu
+
+   The MIT License
+
+   Copyright (c) 2021 Scientific Computing and Imaging Institute,
+   University of Utah.
+
+
+   Permission is hereby granted, free of charge, to any person obtaining a
+   copy of this software and associated documentation files (the "Software"),
+   to deal in the Software without restriction, including without limitation
+   the rights to use, copy, modify, merge, publish, distribute, sublicense,
+   and/or sell copies of the Software, and to permit persons to whom the
+   Software is furnished to do so, subject to the following conditions:
+
+   The above copyright notice and this permission notice shall be included
+   in all copies or substantial portions of the Software.
+
+   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+   OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+   THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+   FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+   DEALINGS IN THE SOFTWARE.
+ */
+
+#ifndef FASTGROWCUT_H
+#define FASTGROWCUT_H
+
+#include <math.h>
+#include <queue>
+#include <set>
+#include <vector>
+#include <stdlib.h>
+#include <fstream>
+#include <iterator>
+
+#include "utilities.h"
+
+#include "fibheap.h"
+#include "HeapNode.h"
+namespace FGC {
+
+const float DIST_INF = std::numeric_limits<float>::max();
+const float DIST_EPSION = 1e-3;
+const unsigned char NNGBH = 26;
+typedef float FPixelType;
+
+template<typename SrcPixelType, typename LabPixelType>
+class FastGrowCut {
+public:
+  FastGrowCut();
+  ~FastGrowCut();
+
+  void SetSourceImage( const std::vector<SrcPixelType>& imSrc );
+  void SetSeedlImage( std::vector<LabPixelType>& imSeed );
+  void SetWorkMode( bool bSegInitialized = false );
+  void SetImageSize( const std::vector<long>& imSize );
+  void DoSegmentation();
+  void GetLabeImage( std::vector<LabPixelType>& imLab );
+  void GetForegroundmage( std::vector<LabPixelType>& imFgrd );
+
+private:
+  void InitializationAHP();
+  void DijkstraBasedClassificationAHP();
+
+  std::vector<SrcPixelType> m_imSrc;
+  std::vector<LabPixelType> m_imSeed;
+  std::vector<LabPixelType> m_imLabPre;
+  std::vector<FPixelType> m_imDistPre;
+  std::vector<LabPixelType> m_imLab;
+  std::vector<FPixelType> m_imDist;
+
+  std::vector<long> m_imSize;
+  long m_DIMX, m_DIMY, m_DIMZ, m_DIMXY, m_DIMXYZ;
+  std::vector<int> m_indOff;
+  std::vector<unsigned char>  m_NBSIZE;
+
+  FibHeap* m_heap;
+  HeapNode* m_hpNodes;
```

Comment:
I know this is ancient code (1996) that you just copied from somewhere, but the memory management of these data structures is unacceptable for 2021 (many raw new/deletes in the client code). I'll checkout the branch and see if I can clean these up.

Reply:
Thank you @dcwhite!

Reply:
Please test this branch, and if it works OK, just merge it into the PR branch
https://github.com/SCIInstitute/Seg3D/tree/growcut-improved-memory

Reply:
Work okay as far as I can tell!